### PR TITLE
Replace NSRecursiveLock with Atomic for observers management in Signal

### PR DESCRIPTION
**Pull request by ikesyo**
It seems that the recursive lock causes deadlocks which is happening at Carthage. But I'm not sure this completely fix the issue. Is this truly right direction?

Might fix https://github.com/Carthage/Carthage/issues/445 and https://github.com/Carthage/Carthage/issues/455.

### CODE REVIEW

**Comment by ikesyo** [comment]
The old `HotSignal` did use a concurrent queue and `dispatch_barrier_sync/async` for observers modification.

File: ReactiveCocoa/Swift/Signal.swift [hunk 1]
```diff
@@ -27,19 +26,17 @@ public final class Signal<T, E: ErrorType> {
...
+			if event.isTerminating {
+				// Disallow any further events (e.g., any triggered recursively).
+				observers = self.observers.swap(nil)
+			} else {
+				observers = self.observers.value
+			}
 
+			if let observers = observers {
 				for sink in observers {
 					sink.put(event)
```
Comment (ikesyo) [inline ReactiveCocoa/Swift/Signal.swift]: The point was putting event to observers outside the lock...

**Comment by norio-nomura** [comment]
As I tested on #455, this branch introduces another crash.
But applying e806194 to v3.0-beta.4 fixed #455 and did not crash.

With that applying, I tested following Cartfile from #445
```
github "ReactiveCocoa/ReactiveCocoa" "swift-development"
github "thoughtbot/Argo" "master"
github "AliSoftware/OHHTTPStubs"
github "danielgindi/ios-charts" "master"

github "Quick/Quick"
github "Quick/Nimble"
```
I got `carthage update` run with very high cpu usage and building didn't progress. : persevere:
It's needed additional fix.

**Comment by jspahrsummers** [comment]
As much as I would like to do this, it begins to violate the contract of `Signal` (one inherited from `RACSignal`), which says that signal events must be serialized. This change could result in events arriving concurrently, and would violate assumptions made in most of our operators right now.

See [the very long previous discussion](https://github.com/ReactiveCocoa/ReactiveCocoa/pull/1120) about recursive locking vs. queue/drain semantics. We ultimately decided that the tradeoffs of recursive locking are preferable.

**Comment by ikesyo** [comment]
Thank you for the reference. :bow: I'll close it out.

**Comment by norio-nomura** [comment]
Oh, I should ask about recursion eight days ago.
I know that original Reactive Extensions using scheduler for preventing recursion, but it did seems RAC did not use scheduler for it. I thought that RAC had different policy…

**Comment by jspahrsummers** [comment]
@norio-nomura Rx does use a queue/drain strategy for schedulers (to avoid a stack overflow from recursive scheduling), but it doesn't extend to Observable events, as far as I know.

**Comment by norio-nomura** [comment]
@jspahrsummers Thank you for answering my stupid question.
Surely Rx uses scheduler on some operators, not on all Observable events.

**Pull request closed without merging**